src/changelog.ts

```diff
@@ -1,16 +1,51 @@
-const semver = require('semver');
+import semver from 'semver';
 
-const { orderedChangeCategories, unreleased } = require('./constants');
+import {
+  changeCategories,
+  orderedChangeCategories,
+  unreleased,
+  Version,
+} from './constants';
 
 const changelogTitle = '# Changelog';
 const changelogDescription = `All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).`;
 
+interface ReleaseMetadata {
+  /**
+   * The version of the current release.
+   */
+  version: Version;
+
+  /**
+   *  An ISO-8601 formatted date, representing the
+   *   release date.
+   */
+  date?: string;
+
+  /**
+   * The status of the release (e.g. 'WITHDRAWN', 'DEPRECATED')
+   */
+  status?: string;
+}
+
+/**
+ * Release changes, organized by category.
+ */
+type ReleaseChanges = Record<changeCategories, string[]>;
```

Comment:
The `ReleaseChanges` objects don't contain _all_ categories, so I think a `Partial` is warranted here

```suggestion
type ReleaseChanges = Partial<Record<changeCategories, string[]>>;
```

This allows getting rid of the couple of `as ReleaseChanges` later on, though not without some cost (TypeScript needs a little extra help to know that we're handling absent categories properly).

Reply:
> ~a little~ _an almost unbelievable amount of_ extra help

FTFY 😅 

Not sure how to do this in a sane way without adding a bunch of postfix non-null assertions (and disabling the related ESLint rule). I'm wondering if there's even a bug in TypeScript. Like, it thinks that `changeDescriptions` in this snippet is `(string | undefined)[]`:

https://github.com/MetaMask/auto-changelog/blob/bcb7beade10c39d57a287feeed75b83ac3a87e83/src/updateChangelog.ts#L72-L83

Reply:
That type seems correct to me. `Partial` _does_ imply that the key can be present and assigned `undefined`. Or it implies that it's possible anyway - either complete absence or an `undefined` assignment. I guess there's no way to distinguish those things with types :thinking:

Reply:
The "TypeScript insanity" changes you made seem pretty reasonable overall. Some of the non-null assertions/casting could be avoided by adding type guards, but it's awkward in literally all of those cases I think. Maybe worth trying anyway, even if it is awkward? But certainly nothing I'd block on.

Reply:
https://github.com/microsoft/TypeScript/pull/43947 is exactly what we need here! In lieu of that, we'll have to go with one ugly or the other.

I tried type guards locally yesterday and I didn't like it any better. I think we might as well stick with this until TypeScript 5 (I guess, since they categorized the above PR as breaking).

Reply:
It's in the 4.4 milestone at the moment. TypeScript doesn't follow semver - they often ship breaking changes in minor releases.

So... maybe relatively soon? :crossed_fingers: